python/tests/integration/test_retry.py

```diff
@@ -0,0 +1,216 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+import pytest
+import tornado
+import tornado.gen
+from mock import patch
+
+from tchannel.errors import ProtocolError
+from tchannel.errors import TChannelError
+from tchannel.errors import TimeoutError
+from tchannel.messages import ErrorCode
+from tchannel.transport_header import RetryType
+from tchannel.tornado import Request
+from tchannel.tornado import TChannel
+from tchannel.tornado.peer import should_retry_on_error
+from tchannel.tornado.stream import InMemStream
+
+
+@pytest.mark.gen_test(timeout=10)
+def test_retry_timeout():
+    endpoint = b'tchannelretrytest'
+
+    @tornado.gen.coroutine
+    def handler_error(request, response, proxy):
+        yield tornado.gen.sleep(2)
+        response.connection.send_error(
+            ErrorCode.busy,
+            "retry",
+            response.id,
+        )
+        # stop normal response streams
+        response.set_exception(TChannelError("stop stream"))
+
+    tchannel = TChannel()
+    tchannel_server = None
+    for i in range(3):
+        tchannel_server = TChannel()
+        tchannel_server.register(endpoint, 'raw', handler_error)
+        tchannel_server.listen()
+        tchannel.peers.get(tchannel_server.hostport)
+
+    hostport = tchannel_server.hostport
+    with (
+        patch(
+            'tchannel.tornado.peer.should_retry_on_error',
+            autospec=True)
+    ) as mock_should_retry_on_error:
+
+        mock_should_retry_on_error.return_value = True
+        with pytest.raises(TimeoutError):
+            yield tchannel.request(
+                hostport,
+                score_threshold=0,
+            ).send(
+                endpoint,
+                "test",
+                "test",
+                headers={
+                    're': RetryType.CONNECTION_ERROR_AND_TIMEOUT
+                },
+                timeout_per_req=0.01,
+                retry_times=3,
+                retry_delay=0.01,
+            )
+
+
+@pytest.mark.gen_test
+def test_retry_on_error_fail():
+    endpoint = b'tchannelretrytest'
+
+    @tornado.gen.coroutine
+    def handler_error(request, response, proxy):
+        response.connection.send_error(
+            ErrorCode.busy,
+            "retry",
+            response.id,
+        )
+
+        # stop normal response streams
+        response.set_exception(TChannelError("stop stream"))
+
+    tchannel = TChannel()
+    tchannel_server = None
+    for i in range(3):
```

Comment:
What? You can put things on the IOLoop outside of a `gen_test` method; inside that method the `IOLoop` will get started and those futures will start being executed. It's still not clear to me why you create the server multiple times, though.